models/lfs.go

```diff
@@ -106,19 +108,88 @@ func (repo *Repository) GetLFSMetaObjectByOid(oid string) (*LFSMetaObject, error
 
 // RemoveLFSMetaObjectByOid removes a LFSMetaObject entry from database by its OID.
 // It may return ErrLFSObjectNotExist or a database error.
-func (repo *Repository) RemoveLFSMetaObjectByOid(oid string) error {
+func (repo *Repository) RemoveLFSMetaObjectByOid(oid string) (int64, error) {
 	if len(oid) == 0 {
-		return ErrLFSObjectNotExist
+		return 0, ErrLFSObjectNotExist
 	}
 
 	sess := x.NewSession()
 	defer sess.Close()
 	if err := sess.Begin(); err != nil {
-		return err
+		return -1, err
 	}
 
 	m := &LFSMetaObject{Oid: oid, RepositoryID: repo.ID}
 	if _, err := sess.Delete(m); err != nil {
+		return -1, err
+	}
+
+	count, err := sess.Count(&LFSMetaObject{Oid: oid})
+	if err != nil {
+		return count, err
+	}
+
+	return count, sess.Commit()
+}
+
+// GetLFSMetaObjects returns all LFSMetaObjects associated with a repository
+func (repo *Repository) GetLFSMetaObjects(page, pageSize int) ([]*LFSMetaObject, error) {
+	sess := x.NewSession()
+	defer sess.Close()
+
+	if page >= 0 && pageSize > 0 {
+		start := 0
+		if page > 0 {
+			start = (page - 1) * pageSize
+		}
+		sess.Limit(pageSize, start)
+	}
+	lfsObjects := make([]*LFSMetaObject, 0, pageSize)
+	return lfsObjects, sess.Find(&lfsObjects, &LFSMetaObject{RepositoryID: repo.ID})
+}
+
+// CountLFSMetaObjects returns a count of all LFSMetaObjects associated with a repository
+func (repo *Repository) CountLFSMetaObjects() (int64, error) {
+	return x.Count(&LFSMetaObject{RepositoryID: repo.ID})
+}
+
+// LFSObjectAccessible checks if a provided Oid is accessible to the user
+func LFSObjectAccessible(user *User, oid string) (bool, error) {
+	if user.IsAdmin {
+		count, err := x.Count(&LFSMetaObject{Oid: oid})
+		return (count > 0), err
+	}
+	cond := accessibleRepositoryCondition(user.ID)
+	count, err := x.Where(cond).Join("INNER", "repository", "`lfs_meta_object`.repository_id = `repository`.id").Count(&LFSMetaObject{Oid: oid})
+	return (count > 0), err
+}
+
+// LFSAutoAssociate auto associates accessible LFSMetaObjects
+func LFSAutoAssociate(metas []*LFSMetaObject, user *User, repoID int64) error {
+	sess := x.NewSession()
+	defer sess.Close()
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+
+	oids := make([]interface{}, len(metas))
+	for i, meta := range metas {
+		oids[i] = meta.Oid
+	}
+
+	cond := builder.NewCond()
+	if !user.IsAdmin {
+		cond = builder.In("`lfs_meta_object`.repository_id",
+			builder.Select("`repository`.id").From("repository").Where(accessibleRepositoryCondition(user.ID)))
+	}
+	newMetas := make([]*LFSMetaObject, 0, len(metas))
+	if err := sess.Cols("oid", "size").Where(cond).In("oid", oids...).GroupBy("oid").Find(&newMetas); err != nil {
```

Comment:
Why is the `size` column needed? A group by just for the `oid` column would resolve directly in the index, but the size column forces a full-scan. And can repeat `oid` values, which I believe will make this fail with duplicate key below.

Reply:
We're only provided with the sha256 oids for the lfs meta objects at the start of the function, not their sizes. We still need the size.

The groupby will only result in duplicates"

Reply:
IIRC we're only provided with the sha256 oids for the lfs meta objects at the start of the function, not their sizes. We still need the size.

The groupby will only result in "duplicates" if there happens to be two objects in the db with the same oid but different sizes. This can only happen if there has been either a sha256 clash or more likely Gitea has had an error somewhere and the size is set to 0. Now, in either case you don't know which is the true size unless you go and check the LFS object dB directly (and check the pointer files.)

Now in this case the autoassociate code is trying to clean up an earlier problem, so I wrote it's fixup to be as expansive as possible.

I'll double check we don't know the size we're looking for but I'm pretty certain we don't have that info.

Reply:
hmm... not sure what to do here.

Reply:
OK I think you're right. We always get the size, so we can just copy it.